blobfuse/OAuthToken.cpp

```diff
@@ -44,60 +51,137 @@ void from_json(const json &j, OAuthToken &t) {
+                    syslog(LOG_WARNING, "Token does not have expires_in date");
+                    expin_failed = true;
+                }
             }
         }
         else
         {
-            printf("no expin");
+            syslog(LOG_WARNING, "Token does not have expires_in date");
             expin_failed = true;
         }
-    } catch(std::exception&){
+    } 
+    catch(std::exception&)
+    {
+        syslog(LOG_WARNING, "Token does not have expires_in date");
         expin_failed = true;
     }
 
-    bool expon_failed = false;
-    try {
-        if (j.contains("expires_on"))
+    if (!expin_failed) // if there is an expires_in just use it and dont worry parsing expires_on field
+    {
+        time_t current_time;
+            
+        struct tm *temptm;
+            
+        time(&current_time);
+            
+        temptm = gmtime( &current_time );
+
+        current_time = mktime(temptm);
+
+        t.expires_on = current_time + t.expires_in;    
+        
+        syslog(LOG_WARNING, "After adding specified expires_in token expiry time in utc %s\n", ctime(&t.expires_on));
+    }
+    else if (j.contains("expires_on"))
+    {
+        try 
         {
             auto val = j.at("expires_on");
             if(val.is_number())
             {
+                // if the expires_on it is localtime. Do not convert it.
+                syslog(LOG_INFO, "expires_on date is a number so using val_get_to to convert it");
                 val.get_to(t.expires_on);
             }
-            else {
+            else 
+            {
                 std::string expires_on = val.get<std::string>();
-                t.expires_on = std::stoi(expires_on);
+                
+                if (is_dt_number(expires_on)) // check with the custom method as the above does not catch everything
+                {
+                    t.expires_on = std::stoi(expires_on);
+                }    
+                else 
+                { // now the date is a string in either the UTC or localtime format so just parse it
+                    fprintf(stdout, "expires_on date is string");
+                    
+                    
+                    // try UTC first then use else to implement in local time in AM or PM
+                    int utcIndex = expires_on.find("+0000");
+                    if ( utcIndex > 19)
+                    { 
+                    // remove tiemzone
+                        expires_on = expires_on.substr(0, utcIndex);
+                        // remove the trailing space if any
+                        expires_on.erase(std::find_if(expires_on.rbegin(), expires_on.rend(), std::bind1st(std::not_equal_to<char>(), ' ')).base(), expires_on.end());
+                        // Now remove the milliseconds because strptime is going crazy with those milliseconds anyway we dont use them so.
+                        int millisecondsIndex  =  expires_on.find(".");
+                        if (millisecondsIndex > 18)

```

Comment:
NIT - Could we add 18 to the constants? (I think this is the size of the time string?)"

Reply:
It is flaky to parse by indexOf, so this is not done perfectly per se, it should be regexed out instead. Noone other our ML is sending string dates anyway. . so I will skip the constant